### PR TITLE
ENH: use numba for kmer processing

### PR DESCRIPTION
**Pull request by rmcar17**
## Summary by Sourcery

Enhancements:
- Optimize kmer processing by using Numba for just-in-time compilation, improving performance.

### CODE REVIEW

File: src/diverse_seq/distance.py [hunk 1]
```diff
@@ -315,13 +324,19 @@ def get_kmers(
...
     -------
-    list[numpy.ndarray]
-        kmers for the sequence.
+    set[int]
+        kmer hashes for the sequence.
     """
-    kmers = []
+    seq = seq.astype(np.int64)
+
+    kmer_hashes = [0]
```
[the remark this answers is not in the document]
Comment (rmcar17) [reply src/diverse_seq/distance.py:337]: Potentially more efficient, though size of array is not known ahead of time.

File: src/diverse_seq/distance.py [hunk 2]
```diff
@@ -333,11 +348,52 @@ def get_kmers(
...
         if i < skip_until:
             continue
-        kmers.append(seq[i : i + k])
-    return kmers
+        kmer_hashes.append(hash_kmer(seq[i : i + k], mash_canonical))
+    return kmer_hashes
+
+
+@numba.njit
+def murmurhash3_32(data: np.ndarray, seed: int = 0x9747B28C) -> int:
```
[the remark this answers is not in the document]
Comment (rmcar17) [reply src/diverse_seq/distance.py:356]: security is not a concern

File: src/diverse_seq/distance.py [hunk 3]
```diff
@@ -351,15 +407,25 @@ def hash_kmer(kmer: np.ndarray, *, mash_canonical: bool) -> int:
     -------
     int
         The has of a kmer.
+
+    Notes
+    -----
+    Uses MurmurHash3 32-bit implementation.
     """
-    tuple_kmer = tuple(map(int, kmer))
+    smallest = kmer
```
[the remark this answers is not in the document]
Comment (rmcar17) [reply src/diverse_seq/distance.py:415]: The two code snippets are not equivalent, they should be ordered lexicographically

File: src/diverse_seq/distance.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rmcar17) [reply src/diverse_seq/distance.py:337]: But at that point, there's no need to convert it to a numpy array, the computation has been done so might as well return the list

**Review by GavinHuttley — APPROVED** [review]
Interesting exchange witha. bot @rmcar17 !

I'm accepting the PR but I pose the following question -- how can you test the implemented hash is correct? If there is a validated implementation elsewhere that we can set a seed, then we could record the results of say 5 random k-mers from that validated implementation and test against those for yours?

**Comment by rmcar17** [comment]
Yep, was curious to see how that worked!

There are implementations elsewhere of that hashing algorithm, adding a few tests along those lines seems like a good idea.

**Pull request merged**